File: PseudoShadow/fipbmp.cs
using ConvolutionLib;
using FloatPointPxBitmap;
using System;
using System.Collections.Generic;
using System.Drawing.Imaging;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PseudoShadow
{
    internal class Fipbmp
    {
        public static bool FastZGaussian_Blur_NxN(Bitmap b, int Length, double Weight, int Sigma, bool doTransparency, 
            bool DoR, bool DoG, bool DoB, bool LeaveNotSelectedChannelsAtCurrentValue, bool SrcOnSigma, Convolution conv, bool doBoth, bool doHorz)
        {
            if ((Length & 0x1) != 1)
                return false;

            double[] KernelVector = new double[Length - 1 + 1];

            int Radius = Length / 2;

            double a = -2.0 * Radius * Radius / Math.Log(Weight);
            double Sum = 0.0;

            for (int x = 0; x <= KernelVector.Length - 1; x++)
            {
                double dist = Math.Abs(x - Radius);
                KernelVector[x] = Math.Exp(-dist * dist / a);
                Sum += KernelVector[x];
            }

            for (int x = 0; x <= KernelVector.Length - 1; x++)
                KernelVector[x] /= Sum;

            double[] AddValVector = conv.CalculateStandardAddVals(KernelVector, Math.Min(255, b.Width - 1));

            ConvolutionLib.ProgressEventArgs pe = new ConvolutionLib.ProgressEventArgs(b.Height * b.Width * 2, 0);

            if (conv.CancelLoops == false)
            {
                if (doBoth || doHorz)
                    conv.ConvolveH_Par(b, KernelVector, AddValVector, 0, Sigma, doTransparency, DoR, DoG, DoB, 
                        LeaveNotSelectedChannelsAtCurrentValue, Math.Min(255, b.Width - 1), SrcOnSigma, pe, 0);
            }

            if (conv.CancelLoops == false)
                b.RotateFlip(RotateFlipType.Rotate270FlipNone);

            if (conv.CancelLoops == false)
            {
                if (doBoth || !doHorz)
                    conv.ConvolveH_Par(b, KernelVector, AddValVector, 0, Sigma, doTransparency, DoR, DoG, DoB, 
                        LeaveNotSelectedChannelsAtCurrentValue, Math.Min(255, b.Width - 1), SrcOnSigma, pe, System.Convert.ToInt32(pe.CurrentProgress));
            }

            if (conv.CancelLoops == false)
                b.RotateFlip(RotateFlipType.Rotate90FlipNone);

            return true;
        }

        public static unsafe void ReplaceColors(Bitmap bmp, int nAlpha, int nRed, int nGreen, int nBlue, int tolerance, int zAlpha, int zRed, int zGreen, int zBlue)
        {
            BitmapData? bmData = null;
            if (AvailMem.AvailMem.checkAvailRam(bmp.Width * bmp.Height * 4L))
            {
                try
                {
                    bmData = bmp.LockBits(new Rectangle(0, 0, bmp.Width, bmp.Height), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);

                    int stride = bmData.Stride;
                    System.IntPtr Scan0 = bmData.Scan0;

                    int nWidth = bmp.Width;
                    int nHeight = bmp.Height;

                    //byte* p = (byte*)bmData.Scan0;// new byte[(bmData.Stride * bmData.Height)];

                    Parallel.For(0, nHeight, y =>
                    {
                        byte* p = (byte*)bmData.Scan0;
                        // for (int y = 0; y < bmp.Height; y++)
                        int pos = 0;
                        pos += y * stride;

                        for (int x = 0; x <= nWidth - 1; x++)
                        {
                            if ((p[pos + 3] == 0 && zAlpha == 0) ||
                                ((p[pos + 3] > (zAlpha - tolerance)) && (p[pos + 3] < (zAlpha + tolerance)) &&
                                (p[pos + 2] > (zRed - tolerance)) && (p[pos + 2] < (zRed + tolerance)) &&
                                (p[pos + 1] > (zGreen - tolerance)) && (p[pos + 1] < (zGreen + tolerance)) &&
                                (p[pos] > (zBlue - tolerance)) && (p[pos] < (zBlue + tolerance))))
                            {
                                int value = nAlpha + System.Convert.ToInt32(p[pos + 3]) - zAlpha;
                                if (value < 0)
                                    value = 0;
                                if (value > 255)
                                    value = 255;
                                p[pos + 3] = System.Convert.ToByte(value);
                                int value2 = nRed + System.Convert.ToInt32(p[pos + 2]) - zRed;
                                if (value2 < 0)
                                    value2 = 0;
                                if (value2 > 255)
                                    value2 = 255;
                                p[pos + 2] = System.Convert.ToByte(value2);
                                int value3 = nGreen + System.Convert.ToInt32(p[pos + 1]) - zGreen;
                                if (value3 < 0)
                                    value3 = 0;
                                if (value3 > 255)
                                    value3 = 255;
                                p[pos + 1] = System.Convert.ToByte(value3);
                                int value4 = nBlue + System.Convert.ToInt32(p[pos]) - zBlue;
                                if (value4 < 0)
                                    value4 = 0;
                                if (value4 > 255)
                                    value4 = 255;
                                p[pos] = System.Convert.ToByte(value4);
                            }

                            pos += 4;
                        }
                    });

                    bmp.UnlockBits(bmData);
                }
                catch
                {
                    try
                    {
                        if (bmData != null)
                            bmp.UnlockBits(bmData);
                    }
                    catch
                    {
                    }
                }
            }
        }
    }
}

File: ColorCurvesAlpha/fipbmp.vb
Option Strict On

Imports System.Runtime.InteropServices
Imports System.Drawing.Imaging
Imports System.Drawing
Imports System.Threading.Tasks

Public Class fipbmp
    Public Shared Sub gradColors(bmp As Bitmap, alphaValues As Integer(), ignoreAlpha0 As Boolean)
        If AvailMem.AvailMem.checkAvailRam(bmp.Width * bmp.Height * 4L) Then
            Dim bmData As BitmapData = Nothing

            Try
                bmData = bmp.LockBits(New Rectangle(0, 0, bmp.Width, bmp.Height), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)

                Dim stride As Integer = bmData.Stride
                Dim Scan0 As System.IntPtr = bmData.Scan0

                Dim nWidth As Integer = bmp.Width
                Dim nHeight As Integer = bmp.Height

                Dim p((bmData.Stride * bmData.Height) - 1) As Byte
                Marshal.Copy(bmData.Scan0, p, 0, p.Length)

                Parallel.[For](0, nHeight, Sub(y)
                                               'For y As Integer = 0 To nHeight - 1
                                               Dim pos As Integer = y * stride

                                               For x As Integer = 0 To nWidth - 1
                                                   If p(pos + 3) > 0 OrElse Not ignoreAlpha0 Then
                                                       Dim value As Integer = alphaValues(p(pos + 3))
                                                       If value < 0 Then
                                                           value = 0
                                                       End If
                                                       If value > 255 Then
                                                           value = 255
                                                       End If
                                                       p(pos + 3) = CByte(value)
                                                   End If
                                                   pos += 4
                                               Next
                                               'Next
                                           End Sub)

                Marshal.Copy(p, 0, bmData.Scan0, p.Length)
                bmp.UnlockBits(bmData)

                p = Nothing
            Catch
                Try
                    bmp.UnlockBits(bmData)
                Catch
                End Try
            End Try
        End If
    End Sub

    Public Shared Sub gradLumToAlpha(bmp As Bitmap, alphaValues As Integer(), doAlpha As Boolean, ignoreAlpha0 As Boolean)
        If AvailMem.AvailMem.checkAvailRam(bmp.Width * bmp.Height * 4L) Then
            Dim bmData As BitmapData = Nothing

            Try
                bmData = bmp.LockBits(New Rectangle(0, 0, bmp.Width, bmp.Height), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)

                Dim stride As Integer = bmData.Stride
                Dim Scan0 As System.IntPtr = bmData.Scan0

                Dim nWidth As Integer = bmp.Width
                Dim nHeight As Integer = bmp.Height

                Dim p((bmData.Stride * bmData.Height) - 1) As Byte
                Marshal.Copy(bmData.Scan0, p, 0, p.Length)

                Parallel.[For](0, nHeight, Sub(y)
                                               'For y As Integer = 0 To nHeight - 1
                                               Dim pos As Integer = y * stride

                                               For x As Integer = 0 To nWidth - 1
                                                   If p(pos + 3) > 0 OrElse Not ignoreAlpha0 Then
                                                       Dim c As Color = Color.FromArgb(p(pos + 3), p(pos + 2), p(pos + 1), p(pos))
                                                       Dim lum As Single = Math.Max(Math.Min(c.GetBrightness() * 255.0F, 255), 0)

                                                       If doAlpha Then
                                                           lum *= CSng(CDbl(p(pos + 3)) / 255.0)
                                                       End If

                                                       Dim value As Integer = alphaValues(CInt(lum))
                                                       If value < 0 Then
                                                           value = 0
                                                       End If
                                                       If value > 255 Then
                                                           value = 255
                                                       End If

                                                       p(pos + 3) = CByte(value)
                                                   End If
                                                   pos += 4
                                               Next
                                               'Next
                                           End Sub)

                Marshal.Copy(p, 0, bmData.Scan0, p.Length)
                bmp.UnlockBits(bmData)

                p = Nothing
            Catch
                Try
                    bmp.UnlockBits(bmData)
                Catch
                End Try
            End Try
        End If
    End Sub

    Public Shared Function ScrollToPic(bmp As Bitmap) As Point
        If AvailMem.AvailMem.checkAvailRam(bmp.Width * bmp.Height * 4L) Then
            ' GDI+ still lies to us - the return format is BGR, NOT RGB.
            Dim bmData As BitmapData = bmp.LockBits(New Rectangle(0, 0, bmp.Width, bmp.Height), ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)

            Dim scanline As Integer = bmData.Stride

            Dim Scan0 As System.IntPtr = bmData.Scan0

            Dim lefttop As New Point()
            Dim complete As Boolean = False

            Dim p((bmData.Stride * bmData.Height) - 1) As Byte
            Marshal.Copy(bmData.Scan0, p, 0, p.Length)

            Dim pos As Integer = 0

            For y As Integer = 0 To bmp.Height - 1
                For x As Integer = 0 To bmp.Width - 1
                    If p(pos + 3) <> 0 Then
                        lefttop = New Point(x, y)
                        complete = True
                        Exit For
                    End If

                    pos += 4
                Next
                If complete Then
                    Exit For
                End If
            Next

            Marshal.Copy(p, 0, bmData.Scan0, p.Length)
            bmp.UnlockBits(bmData)

            p = Nothing

            Return lefttop
        End If

        Return New Point(0, 0)
    End Function
End Class
